{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# *Salix nigra* SLR: Alignments\n",
    "Brian J. Sanderson\n",
    "\n",
    "Last updated: 27 Mar 2020"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Libraries used in this notebook (R 3.6.2)\n",
    "suppressMessages(library(tidyverse))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preparing raw reads for alignments\n",
    "\n",
    "These steps implement the Broad Institute best practices for data pre-processing (current as of 2018-06-29) found [here](https://software.broadinstitute.org/gatk/best-practices/workflow?id=11165). Note that we are not doing base score recalibration because we lack the necessary \"gold standard\" SNP information in our species.\n",
    "\n",
    "## Convert raw FASTQ to uBAM files with appropriate read group info\n",
    "\n",
    "It's possible to work with raw FASTQ reads for alignment, but they lack metadata. The Broad Institute (makers of GATK) recommend converting FASTQ files into unaligned BAM files, which allows you to properly specify the relevant metadata. The following code does this. \n",
    "\n",
    "*Note*: this code chunk, and all of the ones that follow up until the section [Create a scatter file for variant detection](#Create-a-scatter-file-for-variant-detection) below are SLURM submission scripts that were run on the HPCC at Texas Tech University. Some modification is likely necessary to run on different HPC environments.\n",
    "\n",
    "The file ```allnames-short.txt``` referenced below is just a text file with the names of each library, one per line, which allows for the assignment of a prefix variable for the fastq files and the resulting BAM file.\n",
    "\n",
    "```bash\n",
    "# HPC submisison script\n",
    "\n",
    "#!/bin/sh\n",
    "#$ -V\n",
    "#$ -cwd\n",
    "#$ -t 1-96:1\n",
    "#$ -S /bin/bash\n",
    "#$ -N snigra-alignment-qc\n",
    "#$ -q omni\n",
    "#$ -o log/00/$JOB_NAME.o$JOB_ID.$TASK_ID\n",
    "#$ -e log/00/$JOB_NAME.e$JOB_ID.$TASK_ID\n",
    "#$ -pe sm 1\n",
    "#$ -P quanah\n",
    "\n",
    "prefixNum=\"$SGE_TASK_ID\"\n",
    "prefixNumP=\"$prefixNum\"p\n",
    "prefix=`sed -n \"$prefixNumP\" allnames-short.txt`\n",
    "\n",
    "gatk FastqToSam --java-options \"-Xmx8G\" \\\n",
    "    --FASTQ reads/\"$prefix\"_R1_001.fastq \\\n",
    "    --FASTQ2 reads/\"$prefix\"_R2_001.fastq \\\n",
    "    --OUTPUT uBAM/SN\"$prefix\".unmapped.bam \\\n",
    "    --READ_GROUP_NAME Salix_nigra_SLR \\\n",
    "    --SAMPLE_NAME \"$prefix\" \\\n",
    "    --LIBRARY_NAME NEBNext-\"$prefix\" \\\n",
    "    --PLATFORM_UNIT HT23LBBXX.6.1101 \\\n",
    "    --PLATFORM illumina \\\n",
    "    --SEQUENCING_CENTER OMRF \\\n",
    "    --RUN_DATE 2018-03-12T08:55:00+0500\n",
    "```\n",
    "\n",
    "## Mapped the reads to the *Salix purpurea* PacBio v5 assembly\n",
    "\n",
    "Here we map the reads for each library onto the *Salix purpurea* PacBio assembly\n",
    "\n",
    "```bash\n",
    "# HPC submisison script\n",
    "#!/bin/sh\n",
    "#$ -V\n",
    "#$ -cwd\n",
    "#$ -S /bin/bash\n",
    "#$ -t 1-96:1\n",
    "#$ -N snigra-alignment-qc\n",
    "#$ -q omni\n",
    "#$ -e log/00/$JOB_NAME.e$JOB_ID.$TASK_ID\n",
    "#$ -o log/00/$JOB_NAME.o$JOB_ID.$TASK_ID\n",
    "#$ -pe sm 16\n",
    "#$ -P quanah\n",
    "\n",
    "prefixNum=\"$SGE_TASK_ID\"\n",
    "prefixNumP=\"$prefixNum\"p\n",
    "prefix=`sed -n \"$prefixNumP\" allnames.txt`\n",
    "\n",
    "gatk --java-options \"-Dsamjdk.compression_level=5 -Xms3000m\" SamToFastq \\\n",
    "    --INPUT uBAM/\"$prefix\".unmapped.bam \\\n",
    "    --FASTQ /dev/stdout \\\n",
    "    --INTERLEAVE true \\\n",
    "    --NON_PF true \\\n",
    "    | \\\n",
    "    bwa mem -K 100000000 -p -v 3 -t 16 \\\n",
    "    -Y Salix_purpurea_var_94006.mainGenome-noZ.fasta \\\n",
    "    /dev/stdin - \\\n",
    "    | \\\n",
    "    samtools view -1 - > aBAM/\"$prefix\".aligned.bam\n",
    "```\n",
    "\n",
    "## Merge aligned and unaligned BAMs\n",
    "\n",
    "Here we merge the aligned BAMs we created with the previous step with the original unaligned BAM files. This is useful for identifying erroneous reads that resulted from optical duplicate clusters in the next step.\n",
    "\n",
    "```bash\n",
    "# HPC submisison script\n",
    "\n",
    "#!/bin/sh\n",
    "#$ -V\n",
    "#$ -cwd\n",
    "#$ -t 1-96:1\n",
    "#$ -S /bin/bash\n",
    "#$ -N snigra-alignment-qc\n",
    "#$ -q omni\n",
    "#$ -o log/00/$JOB_NAME.o$JOB_ID.$TASK_ID\n",
    "#$ -e log/00/$JOB_NAME.e$JOB_ID.$TASK_ID\n",
    "#$ -pe sm 1\n",
    "#$ -P quanah\n",
    "\n",
    "prefixNum=\"$SGE_TASK_ID\"\n",
    "prefixNumP=\"$prefixNum\"p\n",
    "prefix=`sed -n \"$prefixNumP\" allnames.txt`\n",
    "\n",
    "gatk  --java-options \"-Dsamjdk.compression_level=5 -Xms3000m\" MergeBamAlignment \\\n",
    "    --VALIDATION_STRINGENCY SILENT \\\n",
    "    --EXPECTED_ORIENTATIONS FR \\\n",
    "    --ATTRIBUTES_TO_RETAIN X0 \\\n",
    "    --ALIGNED_BAM aBAM/\"$prefix\".aligned.bam \\\n",
    "    --UNMAPPED_BAM uBAM/\"$prefix\".unmapped.bam \\\n",
    "    --OUTPUT mBAM/\"$prefix\".merged.bam \\\n",
    "    --REFERENCE_SEQUENCE Salix_purpurea_var_94006.mainGenome-noZ.fasta \\\n",
    "    --PAIRED_RUN true \\\n",
    "    --SORT_ORDER 'unsorted' \\\n",
    "    --IS_BISULFITE_SEQUENCE false \\\n",
    "    --ALIGNED_READS_ONLY false \\\n",
    "    --CLIP_ADAPTERS false \\\n",
    "    --MAX_RECORDS_IN_RAM 2000000 \\\n",
    "    --ADD_MATE_CIGAR true \\\n",
    "    --MAX_INSERTIONS_OR_DELETIONS -1 \\\n",
    "    --PRIMARY_ALIGNMENT_STRATEGY MostDistant \\\n",
    "    --PROGRAM_RECORD_ID 'bwamem' \\\n",
    "    --PROGRAM_GROUP_VERSION '0.7.17-r1188' \\\n",
    "    --PROGRAM_GROUP_COMMAND_LINE 'bwa mem -K 100000000 -p -v 3 -t 16 \\\n",
    "        -Y Salix_purpurea_var_94006.mainGenome-noZ.fasta /dev/stdin -' \\\n",
    "    --PROGRAM_GROUP_NAME 'bwamem' \\\n",
    "    --UNMAPPED_READ_STRATEGY COPY_TO_TAG \\\n",
    "    --ALIGNER_PROPER_PAIR_FLAGS true \\\n",
    "    --UNMAP_CONTAMINANT_READS true\n",
    "```\n",
    "\n",
    "## Sort and fix tags\n",
    "\n",
    "Here we sort the alignments by coordinates, and then fix Nm and Uq tags.\n",
    "\n",
    "```bash\n",
    "# HPC submisison script\n",
    "\n",
    "#!/bin/sh\n",
    "#$ -V\n",
    "#$ -cwd\n",
    "#$ -t 1-96:1\n",
    "#$ -S /bin/bash\n",
    "#$ -N snigra-alignment-qc\n",
    "#$ -q omni\n",
    "#$ -o log/00/$JOB_NAME.o$JOB_ID.$TASK_ID\n",
    "#$ -e log/00/$JOB_NAME.e$JOB_ID.$TASK_ID\n",
    "#$ -pe sm 1\n",
    "#$ -P quanah\n",
    "\n",
    "prefixNum=\"$SGE_TASK_ID\"\n",
    "prefixNumP=\"$prefixNum\"p\n",
    "prefix=`sed -n \"$prefixNumP\" allnames.txt`\n",
    "\n",
    "gatk --java-options \"-Dsamjdk.compression_level=5 -Xms4000m\" SortSam \\\n",
    "    --INPUT mBAM/\"$prefix\".merged.bam \\\n",
    "    --OUTPUT /dev/stdout \\\n",
    "    --SORT_ORDER 'coordinate' \\\n",
    "    --CREATE_INDEX false \\\n",
    "    --CREATE_MD5_FILE false \\\n",
    "    | \\\n",
    "    gatk --java-options \"-Dsamjdk.compression_level=5 -Xms500m\" SetNmMdAndUqTags \\\n",
    "    --INPUT /dev/stdin \\\n",
    "    --OUTPUT sBAM/\"$prefix\".sorted.bam \\\n",
    "    --CREATE_INDEX true \\\n",
    "    --CREATE_MD5_FILE true \\\n",
    "    --REFERENCE_SEQUENCE Salix_purpurea_var_94006.mainGenome-noZ.fasta\n",
    "```\n",
    "\n",
    "## Identify duplicated reads using picard\n",
    "\n",
    "Here we identify and remove reads that are erronous results of optical duplicates.\n",
    "\n",
    "```bash\n",
    "# HPC submisison script\n",
    "\n",
    "#!/bin/sh\n",
    "#$ -V\n",
    "#$ -cwd\n",
    "#$ -t 1-96:1\n",
    "#$ -S /bin/bash\n",
    "#$ -N snigra-alignment-qc\n",
    "#$ -q omni\n",
    "#$ -o log/00/$JOB_NAME.o$JOB_ID.$TASK_ID\n",
    "#$ -e log/00/$JOB_NAME.e$JOB_ID.$TASK_ID\n",
    "#$ -pe sm 1\n",
    "#$ -P quanah\n",
    "\n",
    "prefixNum=\"$SGE_TASK_ID\"\n",
    "prefixNumP=\"$prefixNum\"p\n",
    "prefix=`sed -n \"$prefixNumP\" allnames.txt`\n",
    "\n",
    "gatk --java-options \"-Dsamjdk.compression_level=5 -Xms4000m\" MarkDuplicates \\\n",
    "    --INPUT sBAM/\"$prefix\".sorted.bam \\\n",
    "    --OUTPUT dBAM/\"$prefix\".dedup.bam \\\n",
    "    --METRICS_FILE dBAM/\"$prefix\".dedup.metrics \\\n",
    "    --VALIDATION_STRINGENCY SILENT \\\n",
    "    --OPTICAL_DUPLICATE_PIXEL_DISTANCE 2500 \\\n",
    "    --ASSUME_SORT_ORDER 'coordinate' \\\n",
    "    --CREATE_INDEX true \\\n",
    "    --CREATE_MD5_FILE true\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a scatter file for variant detection\n",
    "\n",
    "The following Python code will prepare a set of 100,000 bp intervals across all chromosomes and scaffolds in the *S. purpurea* genome. This will be used in later analyses to parallelize the variant detection\n",
    "\n",
    "```python\n",
    "# Python code\n",
    "import csv\n",
    "from Bio import SeqIO\n",
    "\n",
    "with open(\"Salix_purpurea_var_94006.mainGenome-noZ.fasta\", 'r') as inputSeqFile:\n",
    "    SeqDict = SeqIO.to_dict(SeqIO.parse(inputSeqFile, 'fasta'))\n",
    "\n",
    "   \n",
    "with open(\"salixIntervals.list\", 'wt', newline='\\n') as csvfile:\n",
    "    writer = csv.writer(csvfile, delimiter='\\t')\n",
    "    for chrom in SeqDict:\n",
    "        end = len(SeqDict[chrom].seq)\n",
    "        i = 1\n",
    "        j = 100000\n",
    "        if end < j:\n",
    "            writer.writerow([chrom + \":\" + str(i) + \"-\" + str(end)])\n",
    "            next\n",
    "        else:\n",
    "            while end > j:\n",
    "                writer.writerow([chrom + \":\" + str(i) + \"-\" + str(j)])\n",
    "                if end > j + 100000:\n",
    "                    i = j + 1\n",
    "                    j = j + 100000\n",
    "                    next\n",
    "                elif end < j + 100000:\n",
    "                    i = j + 1\n",
    "                    writer.writerow([chrom + \":\" + str(i) + \"-\" + str(end)])\n",
    "                    j = j + 100000\n",
    "                    break\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quantify depth across the target loci\n",
    "\n",
    "Finally, we use bedtools to quantify the read depth in our analysis-ready BAM files. First, align the bait sequences (```salix_bait_targets.fasta```) to the reference genome\n",
    "\n",
    "```bash\n",
    "bwa mem -K 100000000 -v 3 -t 16 -Y Salix_purpurea_var_94006.mainGenome-noZ.fasta salixSLRtargets.fasta | samtools view -1 - > salix_bait_targets.bam\n",
    "```\n",
    "\n",
    "Next, convert the bait targets and each of our analysis-ready BAM files into BED files. This uses GNU parallel to run this over 20 BAMs at a time.\n",
    "\n",
    "```bash\n",
    "bedtools bamtobed -i salix_bait_targets.bam > salix_bait_targets.bed\n",
    "parallel -j 20 --gnu --eta \"bedtools bamtobed -i {}.dedup.bam > {}.bed\" :::: allnames_short.txt\n",
    "```\n",
    "\n",
    "Finally, use bedtools intersect to quantify the read depth across the targeted loci:\n",
    "\n",
    "```bash\n",
    "parallel -j 20 --gnu --eta \"bedtools intersect -c -a salix_bait_targets.bed \\\n",
    "                            -b {}.bed > {}.depth\" :::: allnames_short.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summarize read depth across targeted loci in R\n",
    "\n",
    "Create a list of the depth files in the current directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "files <- list.files(\".\", pattern = \"*.depth\")\n",
    "big.df <- vector('list', length(files))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a data frame that contains the depths for each library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressMessages(for (i in 1:length(files)){\n",
    "  ## Read table\n",
    "  name_vec <- str_match(files[i], \"(.*)\\\\..*$\")\n",
    "  tmp<-read_delim(name_vec[1], delim = \"\\t\", \n",
    "                                 col_names = c(\"chr\", \"start\", \"stop\", \"gene\", \n",
    "                                               \"score\", \"dir\", \"count\")) %>% \n",
    "                        mutate(., library = name_vec[2])\n",
    "  big.df[[i]] <- tmp\n",
    "\n",
    "})\n",
    "bam_df <- do.call('rbind', big.df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Distribution of read counts across all reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A tibble: 1 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>5%</th><th scope=col>25%</th><th scope=col>50%</th><th scope=col>75%</th><th scope=col>99%</th><th scope=col>MAX</th><th scope=col>mean</th><th scope=col>sd</th><th scope=col>se</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>4</td><td>20</td><td>49</td><td>112</td><td>617</td><td>15834</td><td>44.679</td><td>2.683072</td><td>0.002170975</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 1 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " 5\\% & 25\\% & 50\\% & 75\\% & 99\\% & MAX & mean & sd & se\\\\\n",
       " <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 4 & 20 & 49 & 112 & 617 & 15834 & 44.679 & 2.683072 & 0.002170975\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 1 × 9\n",
       "\n",
       "| 5% &lt;dbl&gt; | 25% &lt;dbl&gt; | 50% &lt;dbl&gt; | 75% &lt;dbl&gt; | 99% &lt;dbl&gt; | MAX &lt;dbl&gt; | mean &lt;dbl&gt; | sd &lt;dbl&gt; | se &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| 4 | 20 | 49 | 112 | 617 | 15834 | 44.679 | 2.683072 | 0.002170975 |\n",
       "\n"
      ],
      "text/plain": [
       "  5% 25% 50% 75% 99% MAX   mean   sd       se         \n",
       "1 4  20  49  112 617 15834 44.679 2.683072 0.002170975"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "group_by(bam_df) %>%\n",
    "  summarise(., \"5%\" = quantile(count, 0.05),\n",
    "               \"25%\" = quantile(count, 0.25),\n",
    "               \"50%\" = quantile(count, 0.5),\n",
    "               \"75%\" = quantile(count, 0.75),\n",
    "               \"99%\" = quantile(count, 0.99),\n",
    "               \"MAX\" = max(count), \n",
    "               \"mean\" = exp(mean(log(count + 1))) - 1,\n",
    "               \"sd\" = exp(sd(log(count + 1))) - 1,\n",
    "               \"se\" = sd / sqrt(n()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Distribution of read counts for each library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A tibble: 48 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>library</th><th scope=col>5%</th><th scope=col>25%</th><th scope=col>50%</th><th scope=col>75%</th><th scope=col>99%</th><th scope=col>MAX</th><th scope=col>mean</th><th scope=col>sd</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>SN502F </td><td> 4</td><td>22</td><td> 51</td><td>110</td><td> 463.0</td><td> 7038</td><td>45.52219</td><td>2.456412</td></tr>\n",
       "\t<tr><td>SN503M </td><td> 2</td><td>11</td><td> 26</td><td> 57</td><td> 256.0</td><td> 1221</td><td>24.10149</td><td>2.295431</td></tr>\n",
       "\t<tr><td>SN504F </td><td> 2</td><td>13</td><td> 29</td><td> 63</td><td> 278.8</td><td> 1830</td><td>26.53666</td><td>2.265926</td></tr>\n",
       "\t<tr><td>SN505F </td><td> 8</td><td>39</td><td> 90</td><td>196</td><td> 881.0</td><td> 7883</td><td>80.57315</td><td>2.601541</td></tr>\n",
       "\t<tr><td>SN506F </td><td> 5</td><td>26</td><td> 60</td><td>130</td><td> 551.8</td><td> 3304</td><td>53.72216</td><td>2.473303</td></tr>\n",
       "\t<tr><td>SN507M </td><td> 8</td><td>37</td><td> 84</td><td>185</td><td> 826.0</td><td> 5701</td><td>76.04291</td><td>2.563148</td></tr>\n",
       "\t<tr><td>SN508F </td><td>10</td><td>46</td><td>106</td><td>230</td><td>1020.8</td><td> 5466</td><td>94.88032</td><td>2.582878</td></tr>\n",
       "\t<tr><td>SN509F </td><td> 4</td><td>19</td><td> 44</td><td> 97</td><td> 443.8</td><td> 3289</td><td>40.34973</td><td>2.447437</td></tr>\n",
       "\t<tr><td>SN510F </td><td> 6</td><td>32</td><td> 74</td><td>164</td><td> 710.0</td><td> 3184</td><td>66.55650</td><td>2.573792</td></tr>\n",
       "\t<tr><td>SN511F </td><td> 3</td><td>16</td><td> 38</td><td> 82</td><td> 358.0</td><td> 1953</td><td>34.37900</td><td>2.373510</td></tr>\n",
       "\t<tr><td>SN512bF</td><td> 2</td><td>13</td><td> 31</td><td> 69</td><td> 299.0</td><td> 1789</td><td>28.66081</td><td>2.305232</td></tr>\n",
       "\t<tr><td>SN512F </td><td> 2</td><td>11</td><td> 32</td><td> 98</td><td>1443.0</td><td>15834</td><td>34.22676</td><td>3.917735</td></tr>\n",
       "\t<tr><td>SN513F </td><td> 7</td><td>31</td><td> 73</td><td>161</td><td> 737.6</td><td> 4519</td><td>66.14159</td><td>2.553659</td></tr>\n",
       "\t<tr><td>SN514F </td><td> 2</td><td>12</td><td> 29</td><td> 63</td><td> 275.0</td><td> 1645</td><td>25.84639</td><td>2.410939</td></tr>\n",
       "\t<tr><td>SN515M </td><td> 4</td><td>19</td><td> 45</td><td> 98</td><td> 420.0</td><td> 3237</td><td>40.38439</td><td>2.431396</td></tr>\n",
       "\t<tr><td>SN516M </td><td> 5</td><td>25</td><td> 57</td><td>123</td><td> 557.8</td><td> 2449</td><td>51.55529</td><td>2.426199</td></tr>\n",
       "\t<tr><td>SN517M </td><td> 2</td><td>13</td><td> 30</td><td> 64</td><td> 278.0</td><td> 1804</td><td>26.89075</td><td>2.323137</td></tr>\n",
       "\t<tr><td>SN518M </td><td> 5</td><td>25</td><td> 57</td><td>125</td><td> 524.0</td><td> 3920</td><td>51.11391</td><td>2.503084</td></tr>\n",
       "\t<tr><td>SN519M </td><td> 4</td><td>19</td><td> 45</td><td> 98</td><td> 424.0</td><td> 2440</td><td>40.59202</td><td>2.411566</td></tr>\n",
       "\t<tr><td>SN520M </td><td> 5</td><td>25</td><td> 57</td><td>124</td><td> 549.0</td><td> 3147</td><td>51.57843</td><td>2.462847</td></tr>\n",
       "\t<tr><td>SN522F </td><td> 5</td><td>24</td><td> 56</td><td>123</td><td> 533.0</td><td> 4054</td><td>50.75176</td><td>2.483804</td></tr>\n",
       "\t<tr><td>SN523F </td><td> 3</td><td>15</td><td> 36</td><td> 79</td><td> 349.0</td><td> 2702</td><td>32.66688</td><td>2.378555</td></tr>\n",
       "\t<tr><td>SN524M </td><td> 6</td><td>27</td><td> 62</td><td>136</td><td> 586.8</td><td> 3507</td><td>56.09886</td><td>2.478292</td></tr>\n",
       "\t<tr><td>SN525F </td><td> 5</td><td>25</td><td> 58</td><td>124</td><td> 538.8</td><td> 3839</td><td>52.01579</td><td>2.426350</td></tr>\n",
       "\t<tr><td>SN526M </td><td> 3</td><td>18</td><td> 43</td><td> 93</td><td> 433.8</td><td> 2889</td><td>38.77670</td><td>2.467687</td></tr>\n",
       "\t<tr><td>SN527M </td><td> 5</td><td>26</td><td> 62</td><td>135</td><td> 587.0</td><td> 4555</td><td>55.48693</td><td>2.533819</td></tr>\n",
       "\t<tr><td>SN528M </td><td> 4</td><td>20</td><td> 46</td><td>101</td><td> 423.0</td><td> 3693</td><td>41.56907</td><td>2.433064</td></tr>\n",
       "\t<tr><td>SN529M </td><td> 2</td><td>15</td><td> 34</td><td> 77</td><td> 342.0</td><td> 2431</td><td>31.49069</td><td>2.440045</td></tr>\n",
       "\t<tr><td>SN530M </td><td> 2</td><td>11</td><td> 26</td><td> 57</td><td> 251.0</td><td> 1535</td><td>23.80657</td><td>2.305677</td></tr>\n",
       "\t<tr><td>SN531F </td><td> 4</td><td>22</td><td> 52</td><td>116</td><td> 515.8</td><td> 3983</td><td>47.02554</td><td>2.528934</td></tr>\n",
       "\t<tr><td>SN533F </td><td> 4</td><td>19</td><td> 44</td><td> 97</td><td> 416.8</td><td> 3044</td><td>40.53856</td><td>2.364615</td></tr>\n",
       "\t<tr><td>SN534F </td><td> 3</td><td>16</td><td> 38</td><td> 82</td><td> 356.0</td><td> 2860</td><td>34.07665</td><td>2.367555</td></tr>\n",
       "\t<tr><td>SN535F </td><td> 5</td><td>25</td><td> 58</td><td>127</td><td> 550.8</td><td> 2828</td><td>52.01157</td><td>2.488886</td></tr>\n",
       "\t<tr><td>SN536F </td><td> 5</td><td>23</td><td> 53</td><td>118</td><td> 515.8</td><td> 3474</td><td>48.72397</td><td>2.450506</td></tr>\n",
       "\t<tr><td>SN537M </td><td>10</td><td>43</td><td> 98</td><td>211</td><td> 924.8</td><td> 4688</td><td>88.45739</td><td>2.503982</td></tr>\n",
       "\t<tr><td>SN538M </td><td>11</td><td>48</td><td>110</td><td>239</td><td>1080.8</td><td> 6232</td><td>99.15361</td><td>2.572052</td></tr>\n",
       "\t<tr><td>SN539F </td><td> 3</td><td>16</td><td> 37</td><td> 82</td><td> 352.0</td><td> 2891</td><td>33.78832</td><td>2.397541</td></tr>\n",
       "\t<tr><td>SN540F </td><td> 3</td><td>19</td><td> 44</td><td> 95</td><td> 419.0</td><td> 4530</td><td>39.26724</td><td>2.452797</td></tr>\n",
       "\t<tr><td>SN541F </td><td> 3</td><td>18</td><td> 42</td><td> 91</td><td> 411.6</td><td> 2301</td><td>37.85548</td><td>2.416970</td></tr>\n",
       "\t<tr><td>SN542M </td><td> 2</td><td>14</td><td> 33</td><td> 73</td><td> 316.0</td><td> 2164</td><td>29.98484</td><td>2.396359</td></tr>\n",
       "\t<tr><td>SN543F </td><td> 7</td><td>33</td><td> 78</td><td>171</td><td> 761.8</td><td> 5648</td><td>70.06631</td><td>2.552557</td></tr>\n",
       "\t<tr><td>SN544M </td><td> 8</td><td>37</td><td> 86</td><td>188</td><td> 840.0</td><td> 5205</td><td>76.97126</td><td>2.594316</td></tr>\n",
       "\t<tr><td>SN545M </td><td> 3</td><td>18</td><td> 43</td><td> 96</td><td> 417.8</td><td> 2461</td><td>39.14329</td><td>2.469702</td></tr>\n",
       "\t<tr><td>SN547M </td><td> 1</td><td>10</td><td> 23</td><td> 51</td><td> 224.0</td><td> 1379</td><td>21.08414</td><td>2.366432</td></tr>\n",
       "\t<tr><td>SN548M </td><td> 4</td><td>21</td><td> 49</td><td>109</td><td> 471.0</td><td> 3346</td><td>44.78764</td><td>2.452692</td></tr>\n",
       "\t<tr><td>SN549M </td><td> 4</td><td>19</td><td> 44</td><td> 96</td><td> 432.0</td><td> 2853</td><td>39.79774</td><td>2.429417</td></tr>\n",
       "\t<tr><td>SN550M </td><td>10</td><td>46</td><td>106</td><td>232</td><td>1048.0</td><td> 5323</td><td>95.82620</td><td>2.568977</td></tr>\n",
       "\t<tr><td>SN551M </td><td> 3</td><td>15</td><td> 36</td><td> 81</td><td> 341.0</td><td> 1713</td><td>32.76568</td><td>2.456619</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 48 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " library & 5\\% & 25\\% & 50\\% & 75\\% & 99\\% & MAX & mean & sd\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t SN502F  &  4 & 22 &  51 & 110 &  463.0 &  7038 & 45.52219 & 2.456412\\\\\n",
       "\t SN503M  &  2 & 11 &  26 &  57 &  256.0 &  1221 & 24.10149 & 2.295431\\\\\n",
       "\t SN504F  &  2 & 13 &  29 &  63 &  278.8 &  1830 & 26.53666 & 2.265926\\\\\n",
       "\t SN505F  &  8 & 39 &  90 & 196 &  881.0 &  7883 & 80.57315 & 2.601541\\\\\n",
       "\t SN506F  &  5 & 26 &  60 & 130 &  551.8 &  3304 & 53.72216 & 2.473303\\\\\n",
       "\t SN507M  &  8 & 37 &  84 & 185 &  826.0 &  5701 & 76.04291 & 2.563148\\\\\n",
       "\t SN508F  & 10 & 46 & 106 & 230 & 1020.8 &  5466 & 94.88032 & 2.582878\\\\\n",
       "\t SN509F  &  4 & 19 &  44 &  97 &  443.8 &  3289 & 40.34973 & 2.447437\\\\\n",
       "\t SN510F  &  6 & 32 &  74 & 164 &  710.0 &  3184 & 66.55650 & 2.573792\\\\\n",
       "\t SN511F  &  3 & 16 &  38 &  82 &  358.0 &  1953 & 34.37900 & 2.373510\\\\\n",
       "\t SN512bF &  2 & 13 &  31 &  69 &  299.0 &  1789 & 28.66081 & 2.305232\\\\\n",
       "\t SN512F  &  2 & 11 &  32 &  98 & 1443.0 & 15834 & 34.22676 & 3.917735\\\\\n",
       "\t SN513F  &  7 & 31 &  73 & 161 &  737.6 &  4519 & 66.14159 & 2.553659\\\\\n",
       "\t SN514F  &  2 & 12 &  29 &  63 &  275.0 &  1645 & 25.84639 & 2.410939\\\\\n",
       "\t SN515M  &  4 & 19 &  45 &  98 &  420.0 &  3237 & 40.38439 & 2.431396\\\\\n",
       "\t SN516M  &  5 & 25 &  57 & 123 &  557.8 &  2449 & 51.55529 & 2.426199\\\\\n",
       "\t SN517M  &  2 & 13 &  30 &  64 &  278.0 &  1804 & 26.89075 & 2.323137\\\\\n",
       "\t SN518M  &  5 & 25 &  57 & 125 &  524.0 &  3920 & 51.11391 & 2.503084\\\\\n",
       "\t SN519M  &  4 & 19 &  45 &  98 &  424.0 &  2440 & 40.59202 & 2.411566\\\\\n",
       "\t SN520M  &  5 & 25 &  57 & 124 &  549.0 &  3147 & 51.57843 & 2.462847\\\\\n",
       "\t SN522F  &  5 & 24 &  56 & 123 &  533.0 &  4054 & 50.75176 & 2.483804\\\\\n",
       "\t SN523F  &  3 & 15 &  36 &  79 &  349.0 &  2702 & 32.66688 & 2.378555\\\\\n",
       "\t SN524M  &  6 & 27 &  62 & 136 &  586.8 &  3507 & 56.09886 & 2.478292\\\\\n",
       "\t SN525F  &  5 & 25 &  58 & 124 &  538.8 &  3839 & 52.01579 & 2.426350\\\\\n",
       "\t SN526M  &  3 & 18 &  43 &  93 &  433.8 &  2889 & 38.77670 & 2.467687\\\\\n",
       "\t SN527M  &  5 & 26 &  62 & 135 &  587.0 &  4555 & 55.48693 & 2.533819\\\\\n",
       "\t SN528M  &  4 & 20 &  46 & 101 &  423.0 &  3693 & 41.56907 & 2.433064\\\\\n",
       "\t SN529M  &  2 & 15 &  34 &  77 &  342.0 &  2431 & 31.49069 & 2.440045\\\\\n",
       "\t SN530M  &  2 & 11 &  26 &  57 &  251.0 &  1535 & 23.80657 & 2.305677\\\\\n",
       "\t SN531F  &  4 & 22 &  52 & 116 &  515.8 &  3983 & 47.02554 & 2.528934\\\\\n",
       "\t SN533F  &  4 & 19 &  44 &  97 &  416.8 &  3044 & 40.53856 & 2.364615\\\\\n",
       "\t SN534F  &  3 & 16 &  38 &  82 &  356.0 &  2860 & 34.07665 & 2.367555\\\\\n",
       "\t SN535F  &  5 & 25 &  58 & 127 &  550.8 &  2828 & 52.01157 & 2.488886\\\\\n",
       "\t SN536F  &  5 & 23 &  53 & 118 &  515.8 &  3474 & 48.72397 & 2.450506\\\\\n",
       "\t SN537M  & 10 & 43 &  98 & 211 &  924.8 &  4688 & 88.45739 & 2.503982\\\\\n",
       "\t SN538M  & 11 & 48 & 110 & 239 & 1080.8 &  6232 & 99.15361 & 2.572052\\\\\n",
       "\t SN539F  &  3 & 16 &  37 &  82 &  352.0 &  2891 & 33.78832 & 2.397541\\\\\n",
       "\t SN540F  &  3 & 19 &  44 &  95 &  419.0 &  4530 & 39.26724 & 2.452797\\\\\n",
       "\t SN541F  &  3 & 18 &  42 &  91 &  411.6 &  2301 & 37.85548 & 2.416970\\\\\n",
       "\t SN542M  &  2 & 14 &  33 &  73 &  316.0 &  2164 & 29.98484 & 2.396359\\\\\n",
       "\t SN543F  &  7 & 33 &  78 & 171 &  761.8 &  5648 & 70.06631 & 2.552557\\\\\n",
       "\t SN544M  &  8 & 37 &  86 & 188 &  840.0 &  5205 & 76.97126 & 2.594316\\\\\n",
       "\t SN545M  &  3 & 18 &  43 &  96 &  417.8 &  2461 & 39.14329 & 2.469702\\\\\n",
       "\t SN547M  &  1 & 10 &  23 &  51 &  224.0 &  1379 & 21.08414 & 2.366432\\\\\n",
       "\t SN548M  &  4 & 21 &  49 & 109 &  471.0 &  3346 & 44.78764 & 2.452692\\\\\n",
       "\t SN549M  &  4 & 19 &  44 &  96 &  432.0 &  2853 & 39.79774 & 2.429417\\\\\n",
       "\t SN550M  & 10 & 46 & 106 & 232 & 1048.0 &  5323 & 95.82620 & 2.568977\\\\\n",
       "\t SN551M  &  3 & 15 &  36 &  81 &  341.0 &  1713 & 32.76568 & 2.456619\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 48 × 9\n",
       "\n",
       "| library &lt;chr&gt; | 5% &lt;dbl&gt; | 25% &lt;dbl&gt; | 50% &lt;dbl&gt; | 75% &lt;dbl&gt; | 99% &lt;dbl&gt; | MAX &lt;dbl&gt; | mean &lt;dbl&gt; | sd &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| SN502F  |  4 | 22 |  51 | 110 |  463.0 |  7038 | 45.52219 | 2.456412 |\n",
       "| SN503M  |  2 | 11 |  26 |  57 |  256.0 |  1221 | 24.10149 | 2.295431 |\n",
       "| SN504F  |  2 | 13 |  29 |  63 |  278.8 |  1830 | 26.53666 | 2.265926 |\n",
       "| SN505F  |  8 | 39 |  90 | 196 |  881.0 |  7883 | 80.57315 | 2.601541 |\n",
       "| SN506F  |  5 | 26 |  60 | 130 |  551.8 |  3304 | 53.72216 | 2.473303 |\n",
       "| SN507M  |  8 | 37 |  84 | 185 |  826.0 |  5701 | 76.04291 | 2.563148 |\n",
       "| SN508F  | 10 | 46 | 106 | 230 | 1020.8 |  5466 | 94.88032 | 2.582878 |\n",
       "| SN509F  |  4 | 19 |  44 |  97 |  443.8 |  3289 | 40.34973 | 2.447437 |\n",
       "| SN510F  |  6 | 32 |  74 | 164 |  710.0 |  3184 | 66.55650 | 2.573792 |\n",
       "| SN511F  |  3 | 16 |  38 |  82 |  358.0 |  1953 | 34.37900 | 2.373510 |\n",
       "| SN512bF |  2 | 13 |  31 |  69 |  299.0 |  1789 | 28.66081 | 2.305232 |\n",
       "| SN512F  |  2 | 11 |  32 |  98 | 1443.0 | 15834 | 34.22676 | 3.917735 |\n",
       "| SN513F  |  7 | 31 |  73 | 161 |  737.6 |  4519 | 66.14159 | 2.553659 |\n",
       "| SN514F  |  2 | 12 |  29 |  63 |  275.0 |  1645 | 25.84639 | 2.410939 |\n",
       "| SN515M  |  4 | 19 |  45 |  98 |  420.0 |  3237 | 40.38439 | 2.431396 |\n",
       "| SN516M  |  5 | 25 |  57 | 123 |  557.8 |  2449 | 51.55529 | 2.426199 |\n",
       "| SN517M  |  2 | 13 |  30 |  64 |  278.0 |  1804 | 26.89075 | 2.323137 |\n",
       "| SN518M  |  5 | 25 |  57 | 125 |  524.0 |  3920 | 51.11391 | 2.503084 |\n",
       "| SN519M  |  4 | 19 |  45 |  98 |  424.0 |  2440 | 40.59202 | 2.411566 |\n",
       "| SN520M  |  5 | 25 |  57 | 124 |  549.0 |  3147 | 51.57843 | 2.462847 |\n",
       "| SN522F  |  5 | 24 |  56 | 123 |  533.0 |  4054 | 50.75176 | 2.483804 |\n",
       "| SN523F  |  3 | 15 |  36 |  79 |  349.0 |  2702 | 32.66688 | 2.378555 |\n",
       "| SN524M  |  6 | 27 |  62 | 136 |  586.8 |  3507 | 56.09886 | 2.478292 |\n",
       "| SN525F  |  5 | 25 |  58 | 124 |  538.8 |  3839 | 52.01579 | 2.426350 |\n",
       "| SN526M  |  3 | 18 |  43 |  93 |  433.8 |  2889 | 38.77670 | 2.467687 |\n",
       "| SN527M  |  5 | 26 |  62 | 135 |  587.0 |  4555 | 55.48693 | 2.533819 |\n",
       "| SN528M  |  4 | 20 |  46 | 101 |  423.0 |  3693 | 41.56907 | 2.433064 |\n",
       "| SN529M  |  2 | 15 |  34 |  77 |  342.0 |  2431 | 31.49069 | 2.440045 |\n",
       "| SN530M  |  2 | 11 |  26 |  57 |  251.0 |  1535 | 23.80657 | 2.305677 |\n",
       "| SN531F  |  4 | 22 |  52 | 116 |  515.8 |  3983 | 47.02554 | 2.528934 |\n",
       "| SN533F  |  4 | 19 |  44 |  97 |  416.8 |  3044 | 40.53856 | 2.364615 |\n",
       "| SN534F  |  3 | 16 |  38 |  82 |  356.0 |  2860 | 34.07665 | 2.367555 |\n",
       "| SN535F  |  5 | 25 |  58 | 127 |  550.8 |  2828 | 52.01157 | 2.488886 |\n",
       "| SN536F  |  5 | 23 |  53 | 118 |  515.8 |  3474 | 48.72397 | 2.450506 |\n",
       "| SN537M  | 10 | 43 |  98 | 211 |  924.8 |  4688 | 88.45739 | 2.503982 |\n",
       "| SN538M  | 11 | 48 | 110 | 239 | 1080.8 |  6232 | 99.15361 | 2.572052 |\n",
       "| SN539F  |  3 | 16 |  37 |  82 |  352.0 |  2891 | 33.78832 | 2.397541 |\n",
       "| SN540F  |  3 | 19 |  44 |  95 |  419.0 |  4530 | 39.26724 | 2.452797 |\n",
       "| SN541F  |  3 | 18 |  42 |  91 |  411.6 |  2301 | 37.85548 | 2.416970 |\n",
       "| SN542M  |  2 | 14 |  33 |  73 |  316.0 |  2164 | 29.98484 | 2.396359 |\n",
       "| SN543F  |  7 | 33 |  78 | 171 |  761.8 |  5648 | 70.06631 | 2.552557 |\n",
       "| SN544M  |  8 | 37 |  86 | 188 |  840.0 |  5205 | 76.97126 | 2.594316 |\n",
       "| SN545M  |  3 | 18 |  43 |  96 |  417.8 |  2461 | 39.14329 | 2.469702 |\n",
       "| SN547M  |  1 | 10 |  23 |  51 |  224.0 |  1379 | 21.08414 | 2.366432 |\n",
       "| SN548M  |  4 | 21 |  49 | 109 |  471.0 |  3346 | 44.78764 | 2.452692 |\n",
       "| SN549M  |  4 | 19 |  44 |  96 |  432.0 |  2853 | 39.79774 | 2.429417 |\n",
       "| SN550M  | 10 | 46 | 106 | 232 | 1048.0 |  5323 | 95.82620 | 2.568977 |\n",
       "| SN551M  |  3 | 15 |  36 |  81 |  341.0 |  1713 | 32.76568 | 2.456619 |\n",
       "\n"
      ],
      "text/plain": [
       "   library 5% 25% 50% 75% 99%    MAX   mean     sd      \n",
       "1  SN502F   4 22   51 110  463.0  7038 45.52219 2.456412\n",
       "2  SN503M   2 11   26  57  256.0  1221 24.10149 2.295431\n",
       "3  SN504F   2 13   29  63  278.8  1830 26.53666 2.265926\n",
       "4  SN505F   8 39   90 196  881.0  7883 80.57315 2.601541\n",
       "5  SN506F   5 26   60 130  551.8  3304 53.72216 2.473303\n",
       "6  SN507M   8 37   84 185  826.0  5701 76.04291 2.563148\n",
       "7  SN508F  10 46  106 230 1020.8  5466 94.88032 2.582878\n",
       "8  SN509F   4 19   44  97  443.8  3289 40.34973 2.447437\n",
       "9  SN510F   6 32   74 164  710.0  3184 66.55650 2.573792\n",
       "10 SN511F   3 16   38  82  358.0  1953 34.37900 2.373510\n",
       "11 SN512bF  2 13   31  69  299.0  1789 28.66081 2.305232\n",
       "12 SN512F   2 11   32  98 1443.0 15834 34.22676 3.917735\n",
       "13 SN513F   7 31   73 161  737.6  4519 66.14159 2.553659\n",
       "14 SN514F   2 12   29  63  275.0  1645 25.84639 2.410939\n",
       "15 SN515M   4 19   45  98  420.0  3237 40.38439 2.431396\n",
       "16 SN516M   5 25   57 123  557.8  2449 51.55529 2.426199\n",
       "17 SN517M   2 13   30  64  278.0  1804 26.89075 2.323137\n",
       "18 SN518M   5 25   57 125  524.0  3920 51.11391 2.503084\n",
       "19 SN519M   4 19   45  98  424.0  2440 40.59202 2.411566\n",
       "20 SN520M   5 25   57 124  549.0  3147 51.57843 2.462847\n",
       "21 SN522F   5 24   56 123  533.0  4054 50.75176 2.483804\n",
       "22 SN523F   3 15   36  79  349.0  2702 32.66688 2.378555\n",
       "23 SN524M   6 27   62 136  586.8  3507 56.09886 2.478292\n",
       "24 SN525F   5 25   58 124  538.8  3839 52.01579 2.426350\n",
       "25 SN526M   3 18   43  93  433.8  2889 38.77670 2.467687\n",
       "26 SN527M   5 26   62 135  587.0  4555 55.48693 2.533819\n",
       "27 SN528M   4 20   46 101  423.0  3693 41.56907 2.433064\n",
       "28 SN529M   2 15   34  77  342.0  2431 31.49069 2.440045\n",
       "29 SN530M   2 11   26  57  251.0  1535 23.80657 2.305677\n",
       "30 SN531F   4 22   52 116  515.8  3983 47.02554 2.528934\n",
       "31 SN533F   4 19   44  97  416.8  3044 40.53856 2.364615\n",
       "32 SN534F   3 16   38  82  356.0  2860 34.07665 2.367555\n",
       "33 SN535F   5 25   58 127  550.8  2828 52.01157 2.488886\n",
       "34 SN536F   5 23   53 118  515.8  3474 48.72397 2.450506\n",
       "35 SN537M  10 43   98 211  924.8  4688 88.45739 2.503982\n",
       "36 SN538M  11 48  110 239 1080.8  6232 99.15361 2.572052\n",
       "37 SN539F   3 16   37  82  352.0  2891 33.78832 2.397541\n",
       "38 SN540F   3 19   44  95  419.0  4530 39.26724 2.452797\n",
       "39 SN541F   3 18   42  91  411.6  2301 37.85548 2.416970\n",
       "40 SN542M   2 14   33  73  316.0  2164 29.98484 2.396359\n",
       "41 SN543F   7 33   78 171  761.8  5648 70.06631 2.552557\n",
       "42 SN544M   8 37   86 188  840.0  5205 76.97126 2.594316\n",
       "43 SN545M   3 18   43  96  417.8  2461 39.14329 2.469702\n",
       "44 SN547M   1 10   23  51  224.0  1379 21.08414 2.366432\n",
       "45 SN548M   4 21   49 109  471.0  3346 44.78764 2.452692\n",
       "46 SN549M   4 19   44  96  432.0  2853 39.79774 2.429417\n",
       "47 SN550M  10 46  106 232 1048.0  5323 95.82620 2.568977\n",
       "48 SN551M   3 15   36  81  341.0  1713 32.76568 2.456619"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "group_by(bam_df, library) %>%\n",
    "  summarise(., \"5%\" = quantile(count, 0.05),\n",
    "               \"25%\" = quantile(count, 0.25),\n",
    "               \"50%\" = quantile(count, 0.5),\n",
    "               \"75%\" = quantile(count, 0.75),\n",
    "               \"99%\" = quantile(count, 0.99),\n",
    "               \"MAX\" = max(count), \n",
    "               \"mean\" = exp(mean(log(count + 1))) - 1,\n",
    "               \"sd\" = exp(sd(log(count + 1))) - 1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 3.6.2",
   "language": "R",
   "name": "ir36"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
